### DISCUSSION TITLE
ngrok loads document root instead of shared site

### DISCUSSION THREAD

**Discussion by dustinatx** [leokhoa/laragon#376] [Q&A] [answered]
I created a ngrok account and added the authtoken to Laragon. I shared a site in Laragon > www > Share. When I load the URL, it shows the Laragon screen - the equivalent of localhost. How can I make it show a specific site?

**Comment by Pen-y-Fan** [comment]
Are you on the latest Laragon 6? You should see the project from the list. Your project needs to be a child of Laragon's www directory.

![Share project](https://user-images.githubusercontent.com/40126936/215323479-4f470f45-a358-4cc2-aa00-ea73df4389bd.png "Share project")

**Comment by dustinatx** [comment]
Yes, I am on Laragon 6. I go to www > Share > and select the project. The ngrok window opens and shows a URL being mapped to that project. 

When I open the ngrok URL, it shows the Laragon (localhost) screen.

**Comment by parish-shrestha** [comment]
Yes, I'm getting the same localhost page that says "Getting Started". I'm on Laragon 6.

**Comment by wirakw** [comment]
I'm getting same error, and when I check the config localhost on C:\laragon\bin\ngrok\conf
![image](https://user-images.githubusercontent.com/31509382/216031649-caeb2735-f5a3-4bbd-aab8-806508fa8631.png)
the localhost  port always change every second.

**Comment by MDARH** [comment]
I'm getting the same problem. I shared organization.test:
![image](https://user-images.githubusercontent.com/19423765/216767034-4375d712-81eb-419f-a506-378c00444652.png)

But it showing me my localhost:
![image](https://user-images.githubusercontent.com/19423765/216767015-132dce8b-2a6d-46d8-8f8c-fbca22c8baef.png)

**Comment by janvalentik** [comment]
I created a shares folder in the laragon folder. In this folder I have bat files that I use to start ngrok.

Content of example bat file:
..\bin\ngrok\ngrok start -config=conf\myproject.yml -region=eu laragon

Content of myproject.yml:
![2023-02-08_084313](https://user-images.githubusercontent.com/659193/217465931-70395584-c2d3-464f-b930-a505c012da17.jpg)


Replace ************* with your autoh token. 
This is how it works for me wher run bat file. It would be great if laragon could generate the part with the tunnel in the yml file.

**Reply by lordjack** [reply]
I try this. But don't work, the file is override all the time

**Comment by jedymatt** [comment]
I'm using Laragon v6 and here's what I did is change the project config in sites-enabled:
```apacheconf
# Asumming your project name is 'project-name'
# Example path: C:\laragon\etc\apache2\sites-enabled\auto.project-name.test.conf

<VirtualHost project-name.test:80>  # changing the wildcard '*' to 'project-name.test'
    DocumentRoot "C:/laragon/www/project-name/public"
    ServerName project-name.test
    ServerAlias *.project-name.test
    <Directory "C:/laragon/www/project-name/public">
        AllowOverride All
        Require all granted
    </Directory>
</VirtualHost>
```

Also, if your project is built using Laravel and encounred a problem that involves not loading the assets, then this might help you. Set your trust proxy to '*':

```php
# Path: <project root>/app/Http/Middleware/TrustProxies.php
class TrustProxies extends Middleware
{
    // ...
    protected $proxies = '*';
    // ...
}
```

**Reply by kidstell** [reply]
perfect fix.

`<VirtualHost project-name.test:80>`

locate `sites-enabled` using this menu path:
Laragon -> Menu -> Apache -> sites-enabled -> select the specific project you are interested in modifying.

A text file would be opened, check for asterix '*' in the first line, do not replace any other asterix symbol.

replace the astetix with your project's local url (project-name.test).

This would not stop your previous setup from working.

Thank you @jedymatt

**Reply by agrinfauzi** [reply]
the best solution, thank you!. everyone please read the comment above

you just need to change from this

```
<VirtualHost *:80>
    DocumentRoot "${ROOT}"
    ServerName ${SITE}
	ServerAlias *.${SITE}
    <Directory "${ROOT}">
        AllowOverride All
        Require all granted
    </Directory>
</VirtualHost>

<VirtualHost *:443>
    DocumentRoot "${ROOT}"
    ServerName ${SITE}
    ServerAlias *.${SITE}
    <Directory "${ROOT}">
        AllowOverride All
        Require all granted
    </Directory>

    SSLEngine on
    SSLCertificateFile      C:/laragon/etc/ssl/laragon.crt
    SSLCertificateKeyFile   C:/laragon/etc/ssl/laragon.key
 
</VirtualHost>
```

to this

```
<VirtualHost ${SITE}:80>*
    DocumentRoot "${ROOT}"
    ServerName ${SITE}
	ServerAlias *.${SITE}
    <Directory "${ROOT}">
        AllowOverride All
        Require all granted
    </Directory>
</VirtualHost>

<VirtualHost ${SITE}:443>
    DocumentRoot "${ROOT}"
    ServerName ${SITE}
    ServerAlias *.${SITE}
    <Directory "${ROOT}">
        AllowOverride All
        Require all granted
    </Directory>

    SSLEngine on
    SSLCertificateFile      C:/laragon/etc/ssl/laragon.crt
    SSLCertificateKeyFile   C:/laragon/etc/ssl/laragon.key
 
</VirtualHost>
```

**Reply by dustinatx** [reply]
I get a 301 Moved Permanently response in the terminal when trying to connect to the ngrok URL

**Comment by josealgardataborges** [comment]
Will this be fixed in Laragon?
We shouldn't be using hacks to fix a software feature.

**Comment by reactmore** [comment]
try manual terminal ngrok http localdomain.test:port

**Reply by josealgardataborges** [reply]
I know how to use ngrok, I just wanted to know if Laragon would be fixed to use it "properly".

**Comment by josealgardataborges** [comment]
@leokhoa Please!!!!

[This is an HUGE security risk, anyone with the shared ngrok link, will see ALL the locally stored projects](https://github.com/leokhoa/laragon/issues/431#issue-1613280091).

**this needs to be fixed ASAP!!!** 

**!!! NO NOT USE THIS UNTIL IT'S FIXED !!!**

**Reply by leokhoa** [reply]
@josealgardataborges : I am not facing the problem and I don't see it is a security risk because if the Ngrok URL doesn't display what you expect, how would you share it?
However, I will make more tests to reproduce and solve the issue.

**Reply by josealgardataborges** [reply]
This is the security Risk... This user shared a "project" folder, instead the "receiver" of the ngrok link, saw his entire www laragon folder

![image](https://user-images.githubusercontent.com/8929655/223444402-d6633d95-4a27-4710-a8ef-9e5d36d69b08.png)

How can't this be a security risk?

For instance, on a laravel instalation you can donwload the .Env files, etc

**Reply by josealgardataborges** [reply]
> @josealgardataborges : I am not facing the problem and I don't see it is a security risk because if the Ngrok URL doesn't display what you expect, how would you share it? However, I will make more tests to reproduce and solve the issue.

THANK YOU!

**Reply by stevenmottiaux** [reply]
Has this issue been fixed yet ?

**Reply by dustinatx** [reply]
It hasn't been fixed in Laragon. I haven't tried all of the suggested workarounds.

**Comment by josealgardataborges** [comment]
I just installed a clean laravel app and shared with ngrok

![4](https://user-images.githubusercontent.com/8929655/223446431-1447a191-b349-405f-b617-d17e1db35156.png)

**Comment by percynguyen92** [comment]
![image](https://user-images.githubusercontent.com/3127083/223892171-8160b3fb-2131-41e5-9166-0f223dd9fcb1.png)
I have same problem
Sửa được không bác dev ơi

**Reply by ainxgans** [reply]
you shoudld use terminal (from laragon) and run this
ngrok http --host-header=custom-domain.test 80

**Reply by josealgardataborges** [reply]
Let's see if @leokhoa can fix "laragon" to do this trick for us, from the menus :) Since "this" command works just fine for me. Everyone should confirm on their side, so that, @leokhoa can implement it on "laragon" codebase.

**Reply by boykioyb** [reply]
@ainxgans oh, it worked, thank you (y)

**Reply by fysioski** [reply]
There is a bug in the software and config. It generates the tunnel just fine but you should add the projectname at the end of the created ngrok url. for example: **https://a888-2a02-a46e-928f-1-79d5-7013-2419-c441.ngrok-free.app** -> http://test.test:80
Should be: **https://a888-2a02-a46e-928f-1-79d5-7013-2419-c441.ngrok-free.app/test** when you want to see the test project. So the laragon software forgets the projectname at the end. Should be easy to fix I guess. For now please add projectname at the end of generated url. Cheers

**Reply by salimbenfarhat** [reply]
thanks you @ainxgans its works for me for simple laragon project but when wordpress is installed i get redirection to my local url "wordpresssite.test" with this message on console : 

"GET /                          301 Moved Permanently"

**Reply by salimbenfarhat** [reply]
I've tried with this command : 
"**ngrok http --host-header=mywordpressshop.lab 80 --scheme http,https**"

and some elements of my page are broken because of https surely?  Locally the page is displayed correctly.

**Comment by goddas** [comment]
This wasn't happening on laragon 5.1. Installed 6.0 a few weeks ago, noticing that in \bin\ngrok\conf my custom configurations get overwritten every time i try to share that site. Pretty much have to use ngrok manually for this reason. Can anyone confirm this behavior?

**Reply by MDARH** [reply]
Yes! The error happening from 6.0, not 5.

**Comment by davidgoweb** [comment, marked as answer]
I found a quick solution to overcome this issue because a few weeks ago, my team also encountered a similar problem.

Run the terminal in Laragon (terminal button) and execute the following command to set manually your key auth from ngrok:
`ngrok config add-authtoken your_token`

Then try running this command to run ngrok according to the local host in the www folder in Laragon:
`ngrok http --host-header=foldername.test 80`

if the host-header doesn't work, try to type the command above manually

Cherrs 🍻

**Reply by dustinatx** [reply]
I get a 301 Moved Permanently response in the terminal when trying to connect to the ngrok URL

**Reply by davidgoweb** [reply]
have you tried to direct access the file? 
like https://xx-xxx.ngrok-free.app/index.php

**Reply by AchmadBudy** [reply]
that's work for me, thanks...

**Comment by reinierbutot** [comment]
None of the above fix this issue for me.

When using Laragon to share i get shown the root folder that contains all of my projects.
When i use ngrok standalone i get the same.
When i add the --host-header=foldername.test it works but i get redirected to foldername.test and the browser says: 'cant open page because the server is not found', because the browser is on an ipad and not the machine hosting the website.

I set C:\Users\Reinier\projects as the Documen Root in Laragon preferences.

Anyone have an idea how to get rid of the redirect to the foldername.test and remain at the https://xx-xxx.ngrok-free.app address so i can view the website?

**Reply by AchmadBudy** [reply]
Try to change app_url on .env with ngrok domain
